core/src/main/java/org/apache/calcite/rel/rules/materialize/MaterializedViewRule.java

```diff
@@ -1036,21 +1037,36 @@ protected NodeLineage generateSwapTableColumnReferencesLineage(
     final Map<RexNode, Integer> exprsLineage = new HashMap<>();
     final Map<RexNode, Integer> exprsLineageLosslessCasts = new HashMap<>();
     for (int i = 0; i < nodeExprs.size(); i++) {
-      final Set<RexNode> s = mq.getExpressionLineage(node, nodeExprs.get(i));
-      if (s == null) {
+      final Set<RexNode> lineages = mq.getExpressionLineage(node, nodeExprs.get(i));
+      if (lineages == null) {
         // Next expression
         continue;
       }
-      // We only support project - filter - join, thus it should map to
-      // a single expression
-      assert s.size() == 1;
-      // Rewrite expr. First we swap the table references following the table
-      // mapping, then we take first element from the corresponding equivalence class
-      final RexNode e = RexUtil.swapTableColumnReferences(rexBuilder,
-          s.iterator().next(), tableMapping, ec.getEquivalenceClassesMap());
-      exprsLineage.put(e, i);
-      if (RexUtil.isLosslessCast(e)) {
-        exprsLineageLosslessCasts.put(((RexCall) e).getOperands().get(0), i);
+      final RexNode expr = nodeExprs.get(i);
+      if (SqlKind.AND == expr.getKind() || SqlKind.OR == expr.getKind()) {
```

Comment:
Why do we need to restrict ourselves to AND/OR expressions? I suppose the same problem can re-appear easily if we have other kind of expressions e.g., `s.time_id + s.time_id`. Can't we apply a generic solution here?

Moreover, I am not sure that `mq.expressionsLineage` works as expected. I checked the javadoc of `RelMdExpressionLineage#createAllPossibleCombinations` and it is supposed to return an equivalent expression with the one given as input. If the input is `A > 1 AND A <3 ` why in the output  we have `A > 1`, `A<3` decomposed. Is this normal?

Reply:
You are right, we can go for a more general solution and support all complex expressions like the one you mentioned. I will also add one or more tests for that.

Regarding `mq.expressionsLineage` I tend to agree with you, it speaks of returning the `set of resulting expressions equivalent to the input expression`, but then it splits it into multiple expressions.

If `mq.expressionsLineage` is broken then maybe we don't have an issue here at all in the first place. I can't see any unit tests for this part of the code, it's hard to tell the original intention behind the method.

Reply:
Finally my diagnosis is that `mq.expressionsLineage` was indeed broken, it was decomposing top `AND` expressions for no apparent good reason.

Luckily there were unit tests, just not where I expected them to be, I have managed to reproduce the issue in one of them, and double checked that the fix did handle it, and at the same time that nothing else broke.

Please have another look @zabetak if you have some spare cycles, thanks a lot for the tip, I was heading down a slippery road :)